### ISSUE TITLE
Empty Layers in Synthesis Generation due to Zero Bias and Input Sample

### ISSUE THREAD

**Issue by ryalberti** [MaximIntegratedAI/ai8x-training#209]
Hello! I’m not sure if this question should be asked in ai8x-training or ai8x-syntheis. Since I am using mostly unedited scripts in synthesis, I believe it has something to do with my training process. The training was done in a fresh install of ai8x-training. 

When I save a sample of my object detection models and use it in quantization in ai8x-synthesis, I receive the below output when using the default . Full output in generation_output_layer_11.txt . 

ERROR: Layer 11 (conv12_2): 8-bit weight quantization supports an output_shift range of [-15, +15]. The specified value of output_shift is -9223372036854775808 which exceeds the system limits. I’ve experienced this in the past with unsimplified training parameters. Layer 11 is always included in this problem, but it may include other layers as well. The value is always negative but changes. 

I see that output_shift is defined in the [training software](https://github.com/MaximIntegratedAI/ai8x-synthesis#output_shift-optional:~:text=By%20default%2C%20output_shift%20is%20generated%20by%20the%20training%20software%2C%20and%20it%20is%20used%20for%20batch%20normalization%20as%20well%20as%20quantization%2Daware%20training.). I have not changed the network file used in the 

By manually setting the output_shift to 0 in layer 11 in the network (svhn-tinierssd_v4), I was able to create the full demo using the attached gen_svhn_human_detection.sh . Below is the full output is also listed in the attached generation_output.txt as well as the full log in generation_log.txt .

`Arranging weights... ---------------------------------------- 100%
Storing weights...   ---------------------------------------- 100%
WARNING: Layer 5: All bias values are zero. Ignoring the input.
WARNING: Layer 6: All bias values are zero. Ignoring the input.
WARNING: Layer 7 (backbone_conv8): All bias values are zero. Ignoring the input.
WARNING: Layer 10 (conv12_1): All bias values are zero. Ignoring the input.
Creating network...  ------ WARNING:--------------------------------- Layer 3: All output values for the given sample input are zero. The generated known-answer test for this network may not be meaningful. See the log file for details.
Creating network...  -------- -------------------------------  20%WARNING: Layer 4: All output values for the given sample input are zero. The generated known-answer test for this network may not be meaningful. See the log file for details.
Creating network...  ---------- -----------------------------  25%WARNING: Layer 5: All output values for the given sample input are zero. The generated known-answer test for this network may not be meaningful. See the log file for details.
Creating network...  ------------ --------------------------- WARNING: 30% Layer 6: All output values for the given sample input are zero. The generated known-answer test for this network may not be meaningful. See the log file for details.
Creating network...  -------------- -------------------------  35%WARNING: Layer 7 (backbone_conv8): All output values for the given sample input are zero. The generated known-answer test for this network may not be meaningful. See the log file for details.
Creating network...  ---------------- -----------------------  40%WARNING: Layer 8 (backbone_conv9): All output values for the given sample input are zero. The generated known-answer test for this network may not be meaningful. See the log file for details.
WARNING: Layer 9 (backbone_conv10): All output values for the given sample input are zero. The generated known-answer test for this network may not be meaningful. See the log file for details.
Creating network...  --------------------WARNING:  Layer 10 (conv12_1): All output values for the given sample input are zero. The generated known-answer test for this network may not be meaningful. See the log file for details.
Creating network...  ------------------------ ---------------  60%WARNING: Layer 13: All output values for the given sample input are zero. The generated known-answer test for this network may not be meaningful. See the log file for details.
WARNING: Layer 14: All output values for the given sample input are zero. The generated known-answer test for this network may not be meaningful. See the log file for details.
WARNING:Creating network...   Layer 17: All output values for the given sample input are zero. The generated known-answer test for this network may not be meaningful. See the log file for details.
---------------------------------- -----  85%WARNING: Layer 18: All output values for the given sample input are zero. The generated known-answer test for this network may not be meaningful. See the log file for details.
Creating network...  ---------------------------------------- 100%`

It looks like this is related to the sample generated from --evaluate and --save-sample in the ai8x-training. In the ai8x-synthesis repo, I can see an [example](https://github.com/MaximIntegratedAI/ai8x-synthesis/issues/250) of creating an object detection demo without these missing layers. My previous demos using my personal database as well as the unchanged SVHN database have produced the same missing layers. I’ve had the same results when training simplified and not simplified versions of the SVHN dataloader with the same data, dataloader, policies, and parameters. 

Do these empty layers have something to do with saving the sample, or something relating to my data or edited files being used to read the data? 

Thank you in advance!
[2022.12.23-165123.zip](https://github.com/MaximIntegratedAI/ai8x-training/files/10363226/2022.12.23-165123.zip)

**Comment by ermanok** [comment]
Hello ryalberti,

We made a few trials with our TinierSSD model and SVHN dataset but could not regenerate the case of `output_shift` outside of the usable range. From your logs, it is seen that you are using a model called `svhn-tinierssd_v4`, which is possible your design. First of all, could you please share that model file with us for further investigation? Also, it seems you are using your dataset so please be sure that your dataloader should satisfy some conditions explained here: https://github.com/MaximIntegratedAI/ai8x-training#data-loader.

**Comment by ryalberti** [comment]
Hi ermanok,

My attached dataloader, svhn_human.py, is essentially svhn.py with 2 changes. In the collate_fn, it allows black and white photos by repeating the black and white tensor into a [3,74,74] rgb tensor. The other major change is in get_bbox to name all the detections an integer 1. Any other changes (IE get_name) are prints for debugging. I believe it does satisfy the conditions you listed. 

For my training model, I use the default ai85net-tinierssd. 

For my first run of synthesis, I used the default svhn-tinierssd.yaml which produced the bad output_shift. I then changed it to include output_shift: 0 in layer 11. Attached is the correct version of my synthesis network human-detection-tinierssd_v5.yaml along with a new log and a new output.  

Thank you!
[emptyLayers.zip](https://github.com/MaximIntegratedAI/ai8x-training/files/10414150/emptyLayers.zip)

**Comment by alicangok** [comment]
Hi ryalberti,

We examined your provided checkpoints and codes in detail, and here are our observations:
- In your provided trained models qat_best and long_qat_checkpoint, which we assume came from two separate training instances, there are certain layers whose weights are all on the order of 1e-40.
   - This indicates major issues with training, the outputs of these layers will effectively return values close to zero independent of the inputs, even without quantization.
   - Such weights are incompatible with MAX7800X hardware carrying out integer operations, because even with the maximum possible output shift of -15, the smallest positive representable floating point number is on the order of 1e-7.
- Without access to your dataset you used, we were unable to reproduce layers with such weights during our trials. When we used your custom dataloader with the original SVHN dataset, we were still able to obtain somewhat respectable mAP results, and the network weights did not lead to underflows.
   - One major thing we can recommend as an example for single class experimentation is the "SVHN_74_simplified" dataset provided in svhn.py in the repo.
   - Please note that if you are going to be making changes to the dataloader, or if you want to use the original SVHN dataset, you will need to delete/move the processed folder containing the pickle files. Otherwise, the __gen_datasets function in svhn.py will not get called, and it will train with the previously generated dataset.
- Furthermore, a learning rate of 0.05 with Adam optimizer is very likely too high to result in any meaningful training, and can lead to numerical instability. Recommended initial values in the literature are usually in the range of 1e-4 to 1e-3.
- While trying to reproduce your problem, we have discovered a minor bug in the quantize.py in ai8x-synthesis where such unusually low weights can lead underflows as you have experienced. Thank you for bringing this to attention, and it will be fixed soon.
   - This being said, correcting this bug for the synthesis will still result in the networks you provided produce trivial outputs, since the problematic layers will yield quantized weights of exactly 0, even at output_shift = -15.

All the best and good luck!

**Comment by ryalberti** [comment]
Thank you so much, very helpful and very appreciated! 
Sorry for the delayed response, I wasn't working on this project for a few months.
